File: cloud/sdk/py/src/mcp_monitor_sdk/retry.py
import asyncio

from .logger import Logger


async def with_retry(
    fn,
    *,
    max_attempts: int,
    initial_delay: float,
    max_delay: float,
    backoff_multiplier: float,
    logger: Logger,
):
    attempts = max(max_attempts, 1)
    delay = initial_delay
    last_error: Exception | None = None

    for attempt in range(1, attempts + 1):
        try:
            return await fn()
        except Exception as error:  # noqa: BLE001
            last_error = error
            if attempt == attempts:
                logger.error("Retry attempts exhausted", error, attempts=attempt)
                break
            logger.warn(
                "Operation failed, retrying",
                attempt=attempt,
                max_attempts=attempts,
                delay=delay,
                error=str(error),
            )
            await asyncio.sleep(delay)
            delay = min(delay * backoff_multiplier, max_delay)

    assert last_error is not None
    raise last_error

File: cloud/sdk/src/utils/retry.ts
import type { Logger } from "../logger/logger.js";

export interface RetryOptions {
  maxAttempts: number;
  initialDelay: number;
  maxDelay: number;
  backoffMultiplier: number;
}

export async function withRetry<T>(
  fn: () => Promise<T>,
  options: RetryOptions,
  logger: Logger
): Promise<T> {
  let lastError: Error | undefined;
  let delay = options.initialDelay;

  for (let attempt = 1; attempt <= options.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error as Error;

      if (attempt === options.maxAttempts) {
        logger.error("Retry attempts exhausted", lastError, {
          attempts: attempt,
          maxAttempts: options.maxAttempts,
        });
        break;
      }

      logger.warn("Operation failed, retrying", {
        attempt,
        maxAttempts: options.maxAttempts,
        delay,
        error: lastError.message,
      });

      await sleep(delay);

      delay = Math.min(delay * options.backoffMultiplier, options.maxDelay);
    }
  }

  throw lastError;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
